File: 1888-minimum-number-of-flips-to-make-the-binary-string-alternating/1888-minimum-number-of-flips-to-make-the-binary-string-alternating.cpp
class Solution {
public:
    int minFlips(string s) {
        int n = s.size();
        s += s;
        string s1, s2;
        for(int i=0; i < s.size(); i++) {
            s1.push_back(i % 2 ? '0' : '1');
            s2.push_back(i % 2 ? '1' : '0');
        }
        
        int res1 = 0, res2 = 0, res = INT_MAX;
        for(int i=0; i < s.size(); i++) {
            if(s[i] != s1[i])
                res1++;
            if(s[i] != s2[i])
                res2++;
            if(i >= n) {
                if(s[i-n] != s1[i-n])
                    res1--;
                if(s[i-n] != s2[i-n])
                    res2--;
            }
            if(i >= n-1)
                res = min({res, res1, res2});
        }
        return res;
    }
};

File: 1888-minimum-number-of-flips-to-make-the-binary-string-alternating/1888-minimum-number-of-flips-to-make-the-binary-string-alternating.go
func minFlips(s string) int {
    n := len(s)
    s = s + s

    alt1 := make([]byte, len(s))
    alt2 := make([]byte, len(s))

    for i := range s {
        if i%2 == 0 {
            alt1[i] = '0'
            alt2[i] = '1'
        } else {
            alt1[i] = '1'
            alt2[i] = '0'
        }
    }

    res := n
    diff1, diff2 := 0, 0
    l := 0

    for r := 0; r < len(s); r++ {
        if s[r] != alt1[r] { diff1++ }
        if s[r] != alt2[r] { diff2++ }

        if r-l+1 > n {
            if s[l] != alt1[l] { diff1-- }
            if s[l] != alt2[l] { diff2-- }
            l++
        }

        if r-l+1 == n {
            res = min(res, diff1, diff2)
        }
    }

    return res
}